### ISSUE TITLE
Support custom scripts/hooks

### ISSUE THREAD

**Issue by edongashi** [hoarder-app/hoarder#638]
### Describe the feature you'd like

Support hooking custom user code into the bookmark process. A `/scripts` directory could be mounted where named callbacks would be exported. For example, `my-script.js` would expose:

```js
// --- Pseudocode for illustration purposes ---
module.exports = {
  async onBookmark(input, bookmark) {
    if (input.url && input.url.host.contains('youtube.com')) {
      bookmark.tags.add('videos');
    }
    if (input.url && input.url.host.contains('reddit.com')) {
      if (input.url.path.startswith('/r/funny') {
        bookmark.tags.add('entertainment')
      } else if (input.url.path.startswith('/r/programming') {
        bookmark.tags.add('technology')
        bookmark.tags.add('learning')
      }
    }
    return bookmark
  }
}
```

Similarly, we could have a `onAITags` to clean up the AI generated tags, `onAiPrompt` to fine tune a prompt, `onContentCaptured`, `onSummary`, and so on.

### Describe the benefits this would bring to existing Hoarder users

There are some key sites where I commonly save links and I would like to apply some pre/postprocessing. Automatic tagging would be beneficial in a site such as reddit where it depends on which subreddit the bookmark was captured. This would also make it possible for third party integrations such as caling a webhook when something of interest has happened.

Another use case that comes to mind is parsing domain-specific languages for text notes.

### Can the goal of this request already be achieved via other means?

No, the closest thing is customizing the AI prompts.

### Have you searched for an existing open/closed issue?

- [X] I have searched for existing issues and none cover my fundamental request

**Comment by treardon17** [comment]
I would also like this feature. The AI generated tags tend to have inconsistent casing which lead to different iterations of the same tag multiple times.

**Comment by maelp** [comment]
Interesting, although to avoid cluttering Karakeep with feature creep, perhaps the best would be just to setup webhooks when a link is stored, etc, and then you could have your own server / n8n / val.town script which does the post-processing?

**Comment by edongashi** [comment]
> Interesting, although to avoid cluttering Karakeep with feature creep, perhaps the best would be just to setup webhooks when a link is stored, etc, and then you could have your own server / n8n / val.town script which does the post-processing?

Yes, the webhooks feature has some overlap with this one. Problem is that I would need to host and maintain an additional service that does the processing. Sometimes I just want to clean up the link a little, for example remove sharing tokens from URL params and so on. It feels a little cumbersome to do a back and forth for such simple logic. But in any case I might look into running a webhook service with some basic abstractions.

**Comment by edongashi** [comment]
I see the latest update has brought GUI "flows", which is pretty close to what I need but not quite. I need  to rewrite URLs to remove those pesky sharing tokens, and some sites also need a little tweaking to normalize the URL...
